cmd/proxy/wsl_integration_linux.go

```diff
@@ -34,44 +33,80 @@ var (
 )
 
 const (
-	k8sAPI            = "https://192.168.1.2:6443"
+	k8sAPI            = "192.168.1.2:6443"
 	defaultListenAddr = "127.0.0.1:6443"
 )
 
 func main() {
-	flag.BoolVar(&debug, "debug", false, "enable additional debugging")
-	flag.StringVar(&upstreamAddr, "upstream-addr", k8sAPI, "The upstream server's address.")
+	flag.BoolVar(&debug, "debug", false, "enable additional debugging.")
+	flag.StringVar(&upstreamAddr, "upstream-addr", k8sAPI, "The upstream server's address (k3s API sever).")
 	flag.StringVar(&listenAddr, "listen-addr", defaultListenAddr, "The server's address in an IP:PORT format.")
 	flag.Parse()
 
 	if debug {
 		logrus.SetLevel(logrus.DebugLevel)
 	}
 
-	targetURL, err := url.Parse(upstreamAddr)
+	listener, err := net.Listen("tcp", listenAddr)
 	if err != nil {
-		logrus.Fatalf("invalid upstream URL: %s", upstreamAddr)
+		logrus.Fatalf("Failed to listen on %s: %s", listenAddr, err)
 	}
 
-	ctx, _ := signal.NotifyContext(context.Background(), syscall.SIGINT, syscall.SIGTERM)
-	proxy := httputil.NewSingleHostReverseProxy(targetURL)
+	// Handle graceful shutdown
+	sigCh := make(chan os.Signal, 1)
+	signal.Notify(sigCh, syscall.SIGINT, syscall.SIGTERM)
+
+	// WaitGroup to wait for all connections to finish before shutting down
+	var wg sync.WaitGroup
+
+	go func() {
+		<-sigCh
+		logrus.Println("Shutting down...")
+		listener.Close()
+		wg.Wait()
+		os.Exit(0)
+	}()
+
+	logrus.Infof("Proxy server started listening on %s, forwarding to %s", listenAddr, upstreamAddr)
 
-	srv := http.Server{
-		Addr:              listenAddr,
-		Handler:           proxy,
-		ReadHeaderTimeout: 5 * time.Second,
+	for {
+		conn, err := listener.Accept()
+		if err != nil {
+			// Check if the error is due to listener being closed
+			if opErr, ok := err.(*net.OpError); ok && opErr.Err.Error() == "use of closed network connection" {
```

Comment:
Use `errors.As` instead of the cast? Or even better, `errors.Is(net.ErrClosed)` (see [docs](https://pkg.go.dev/net#ErrClosed)).